tests/two_d_array_ut/two_d_array_ut.c

```diff
@@ -0,0 +1,642 @@
+    }
+
+    //act
+    for (unsigned int i = 0; i < 5; i++)
+    {
+        result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(tdarr, i);
+        ASSERT_ARE_EQUAL(int, result, 0);
+        ASSERT_IS_NOT_NULL(tdarr->row_arrays[i]);
+    }
+
+    //assert
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_017: [ Otherwise, TWO_D_ARRAY_ALLOCATE_NEW_ROW(T) shall allocate memory for the new row and return zero on success. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_ALLOCATE_NEW_ROW_of_THANDLE_succeeds)
+{
+    //arrange
+    int result;
+    TWO_D_ARRAY(THANDLE(A_TEST)) tdarr = TWO_D_ARRAY_CREATE(THANDLE(A_TEST))(5, 5000000);
+    umock_c_reset_all_calls();
+    STRICT_EXPECTED_CALL(malloc_2(5000000, sizeof(THANDLE(A_TEST))));
+
+    //act
+    result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(THANDLE(A_TEST))(tdarr, 0);
+
+    //assert
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(int, result, 0);
+    ASSERT_IS_NOT_NULL(tdarr->row_arrays[0]);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    THANDLE(A_TEST) a = THANDLE_MALLOC(A_TEST)(NULL);
+    ASSERT_IS_NOT_NULL(a);
+
+    THANDLE_INITIALIZE(A_TEST)(&tdarr->row_arrays[0][0], a);
+    THANDLE_ASSIGN(A_TEST)(&tdarr->row_arrays[0][0], NULL);
+    THANDLE_ASSIGN(A_TEST)(&a, NULL);
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(THANDLE(A_TEST))(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_014: [ If two_d_array is NULL, TWO_D_ARRAY_ALLOCATE_NEW_ROW(T) shall fail and return a non-zero value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_ALLOCATE_NEW_ROW_fails_when_handle_is_null)
+{
+    //arrange
+    int result;
+
+    //act
+    result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(NULL, 5);
+
+    //assert
+    ASSERT_ARE_NOT_EQUAL(int, result, 0);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_015: [ If row_index is equal or greater than row_size, TWO_D_ARRAY_ALLOCATE_NEW_ROW(T) shall fail and return a non-zero value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_ALLOCATE_NEW_ROW_fails_when_row_index_exceeds)
+{
+    //arrange
+    int result;
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
+    umock_c_reset_all_calls();
+
+    //act
+    result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(tdarr, 5);
+
+    //assert
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_NOT_EQUAL(int, result, 0);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_016: [ If the row specified by row_index has already been allocated, TWO_D_ARRAY_ALLOCATE_NEW_ROW(T) shall fail and return a non-zero value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_ALLOCATE_NEW_ROW_fails_when_that_row_has_already_been_added)
```

Comment:
>added [](http://example.com/codeflow?start=80&length=5)

I think add is misleading, allocated is the correct term me thinks.
add usually implies that anything existing stays the same and something is added to the collection.

Reply:
changed to allocated